### PR TITLE
refactor: introducing `hyperswitch_interface` crates

### PR DESCRIPTION
**Pull request by Chethan-rao**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Create a new crate `Hyperswitch-interface` containing below modules for now: 
`secrets_interface`: Module used for application config `encryption` and `decryption` during startup which will contain its own trait and error types
`encryption_interface`: Module used for user value `encryption` and `decryption` during runtime which will contain its own trait and error types

This PR only adds the above crate with the addition of `secret_management_config` and `encryption_management_config` in the `router`

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

New crates addition which doesn't integrate with other existing crates yet. So basic sanity testing should suffice

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/drainer/src/connection.rs [hunk 1]
```diff
@@ -28,8 +31,8 @@ pub async fn redis_connection(
 pub async fn diesel_make_pg_pool(
     database: &Database,
     _test_transaction: bool,
-    #[cfg(feature = "aws_kms")] aws_kms_client: &'static aws_kms::AwsKmsClient,
-    #[cfg(feature = "hashicorp-vault")] hashicorp_client: &'static hashicorp_vault::HashiCorpVault,
+    #[cfg(feature = "aws_kms")] aws_kms_client: &'static aws_kms::core::AwsKmsClient,
```
Comment (jarnura) [inline crates/drainer/src/connection.rs:34]: why we need both clients?

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/drainer/src/connection.rs (hunk 1 above)
Comment (Chethan-rao) [reply crates/drainer/src/connection.rs:34]: I haven't incorporated my new interface changes in binaries yet. This PR is solely for the interface creation and all the client generation

**Review by jarnura — DISMISSED** [review]

File: crates/external_services/src/aws_kms/decrypt.rs [hunk 2]
```diff
@@ -1,6 +1,7 @@
+//! decrypting data using the AWS KMS SDK.
```
Comment (SanchithHegde) [inline crates/external_services/src/aws_kms/decrypt.rs:1]: Nit:
```suggestion
//! Decrypting data using the AWS KMS SDK.
```

File: crates/external_services/src/aws_kms/decrypt.rs [hunk 3]
```diff
@@ -1,6 +1,7 @@
...
 use common_utils::errors::CustomResult;
 
-use super::*;
+use crate::aws_kms::core::{AwsKmsClient, AwsKmsError, AWS_KMS_CLIENT};
 
 #[async_trait::async_trait]
 /// This trait performs in place decryption of the structure on which this is implemented
```
Comment (SanchithHegde) [inline crates/external_services/src/aws_kms/decrypt.rs:7]: I believe we'll be eventually removing this trait?

File: crates/external_services/src/hashicorp_vault/core.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
+//! Interactions with the HashiCorp Vault
+
+use std::{collections::HashMap, future::Future, pin::Pin};
+
+use common_utils::{ext_traits::ConfigExt, fp_utils::when};
+use error_stack::{Report, ResultExt};
+use masking::Secret;
+use vaultrs::client::{VaultClient, VaultClientSettingsBuilder};
+
+static HC_CLIENT: tokio::sync::OnceCell<HashiCorpVault> = tokio::sync::OnceCell::const_new();
```
Comment (SanchithHegde) [inline crates/external_services/src/hashicorp_vault/core.rs:10]: We can use `once_cell::sync::OnceCell` here, and make `get_hashicorp_client()` a sync function.

File: crates/external_services/src/hashicorp_vault/core.rs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+/// Configuration for connecting to HashiCorp Vault.
+#[derive(Clone, Debug, Default, serde::Deserialize)]
+#[serde(default)]
+pub struct HashiCorpVaultConfig {
+    /// The URL of the HashiCorp Vault server.
+    pub url: String,
+    /// The authentication token used to access HashiCorp Vault.
+    pub token: String,
```
Comment (SanchithHegde) [inline crates/external_services/src/hashicorp_vault/core.rs:26]: We can make this a `Secret`.

File: crates/external_services/src/no_encryption/implementers.rs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+#[async_trait::async_trait]
+impl SecretManagementInterface for NoEncryption {
+    async fn get_secret(
+        &self,
+        input: Secret<String>,
+    ) -> CustomResult<Secret<String>, SecretsManagementError> {
+        String::from_utf8(self.decrypt(input.expose()))
+            .map(Into::into)
+            .into_report()
+            .change_context(SecretsManagementError::FetchSecretFailed)
```
Comment (SanchithHegde) [inline crates/external_services/src/no_encryption/implementers.rs:33]: Maybe add context that the input wasn't a valid (UTF-8) string?

File: crates/external_services/src/aws_kms/decrypt.rs (hunk 3 above)
Comment (Chethan-rao) [reply crates/external_services/src/aws_kms/decrypt.rs:7]: Yes, will take this up in next PR

File: crates/external_services/src/hashicorp_vault/core.rs (hunk 4 above)
Comment (Chethan-rao) [reply crates/external_services/src/hashicorp_vault/core.rs:10]: Same here, will take this up in next PR

**Review by SanchithHegde — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**